### ISSUE TITLE
PeerPool / Discovery don't answer event bus requests after some time

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/py-evm#1583]
### What is wrong?

Yesterday I put #1556 to work on a server ant today I noticed that the node has more or less stalled with the log being overwhelmed by these warnings.

```
 WARNING  12-12 08:36:55       EthstatsService  Timeout: PeerPool did not answer PeerCountRequest
 WARNING  12-12 08:36:56           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
 WARNING  12-12 08:36:59           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
 WARNING  12-12 08:37:00       EthstatsService  Timeout: PeerPool did not answer PeerCountRequest
 WARNING  12-12 08:37:01           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
 WARNING  12-12 08:37:04           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
 WARNING  12-12 08:37:06       EthstatsService  Timeout: PeerPool did not answer PeerCountRequest
 WARNING  12-12 08:37:06           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
 WARNING  12-12 08:37:09           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
 WARNING  12-12 08:37:11           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
 WARNING  12-12 08:37:11       EthstatsService  Timeout: PeerPool did not answer PeerCountRequest
 WARNING  12-12 08:37:14           ETHPeerPool  Discovery did not answer PeerCandidateRequest in time
```

The interesting thing here is, I noticed this one before:

>EthstatsService  Timeout: PeerPool did not answer PeerCountRequest

And so far, I thought this may just be the `PeerPool` being to busy to answer the request by the `EthstatsService` timely.

But now after seeing this one, I feel like I need to look deeper into this.

>ETHPeerPool  Discovery did not answer PeerCandidateRequest in time

I also ran `trinity attach` to see if `w3.net.peerCount` would be responding and in fact it times out. The same happens with `w3.net.version` which isn't even served from the PeerPool (but it is still running on the same event loop / event bus endpoint)

This leads to the question what if there is some form of memory leak / bug in lahja where after some time, the event bus just stalls and cant process any further messages.

### How can it be fixed

There are two things I plan to do:

1. I will take the existing lahja benchmark and start it with parameters that will effectively cause a very long run over several hours to see if they stall at some point. This can give us an indicator if there's something wrong with lahja.

**Update**: Started this now: `python scripts/perf_benchmark.py --num-processes 10 --num-events 36000 --throttle 0.5 &> benchmark2.log`

This should push broadcast one event every 0.5 seconds, to 10 connected processes, 36000 events in total. If my math is correct, this would run over the next 5 hours.

2. I will write a tiny stupid plugin to just ping pong messages via the EventBus that I will run in the context of Trinity for several hours to see if that stalls as well. This can give us an indicator if there's something wrong with how the event bus is integrated in Trinity or it may also show us that there's no problem with the EventBus but with the PeerPool / Discovery instead.

3. Looking through the logs to find anything suspicious around the time when it stops responding to requests.

**Comment by cburgdorf** [comment]
@pipermerriam @carver This is the kind of stuff that still scares me about the EventBus. As much as I love it and the architecture implications it brings to the table. It may still be full of unknown bugs at this point :sweat_smile:

**Comment by carver** [comment]
> I have to admit though, that I've seen it before (especially before #1556) that our peer count would drop to zero after a while so I may be totally off here to blame the EventBus.

Yeah, that would be reasonable (and there are likely undiagnosed problems there).

But this seems like a pretty strong indicator of an event bus problem:

> I also ran `trinity attach` to see if `w3.net.peerCount` would be responding and in fact it times out. The same happens with `w3.net.version` which isn't even served from the PeerPool (but it is still running on the same event loop / event bus endpoint)

**Comment by cburgdorf** [comment]
I have done various long running tests but I wasn't able to reproduce any of the issues. My best bet right now is that there may be something in our `networking` process that can lead to a stalled event loop in general (which could explain the described symptoms) but there's no strong evidence that this would be caused by the `EventBus`. That doesn't mean it isn't, just that I couldn't find strong evidence for it.

I will close this for now.

**Comment by pipermerriam** [comment]
I'm re-opening as this is has occurred around 3 times in the last few days and I'm pretty sure it has always been accompanied by https://github.com/ethereum/py-evm/issues/1613 during shutdown.

**Comment by cburgdorf** [comment]
Meh. Too bad. I'll check if I can reproduce #1613 in isolation.
